go/pkg/pwapi/api_agent-update-state.go

```diff
@@ -15,17 +16,35 @@ func (svc *service) AgentUpdateState(ctx context.Context, in *AgentUpdateState_I
 		return nil, errcode.ErrMissingInput
 	}
 
-	for _, challengeInstance := range in.Instances {
-		cpy := challengeInstance
-		err := svc.db.Model(&cpy).
-			Update(pwdb.ChallengeInstance{
-				Status:         challengeInstance.Status,
-				InstanceConfig: challengeInstance.InstanceConfig,
-			}).
-			Error
-		if err != nil {
-			return nil, errcode.ErrAgentUpdateState.Wrap(err)
+	userID, err := userIDFromContext(ctx, svc.db)
+	if err != nil {
+		return nil, errcode.ErrGetUserIDFromContext.Wrap(err)
+	}
+
+	err = svc.db.Transaction(func(tx *gorm.DB) error {
+		for _, challengeInstance := range in.Instances {
+			cpy := challengeInstance
+			err := tx.Model(&cpy).
+				Update(pwdb.ChallengeInstance{
+					Status:         challengeInstance.Status,
+					InstanceConfig: challengeInstance.InstanceConfig,
+				}).
+				Error
+			if err != nil {
+				return errcode.ErrAgentUpdateState.Wrap(err)
+			}
+		}
+		activity := pwdb.Activity{
+			AuthorID: userID,
+			Kind:     pwdb.Activity_AgentUpdate,
+		}
```

Comment:
this activity is missing context, you should attach as much info as you can -> challengeinstance id, challenge flavor id if you have it, etc